### PR TITLE
Regions rework

### PR DESCRIPTION
**Pull request by mikbras**
This change is a total rework of memory regions. The main goal is to minimize changes to OE by moving region demarcation from OE to Mystikos.

The OE-side design is described here:

https://github.com/openenclave/openenclave/blob/ce020e8841044cd9796d288e64d4e5205df1130c/docs/DesignDocs/extra_enclave_data.md

Each region holds a particular unit such as:

- the kernel
- the C runtime
- the CPIO root file system
- the memory management pages
- the archive (containing roothashes and public keys)

These regions are flattened into a contiguous memory image. The Linux implementation and SGX now share much of the implementation, which simplifies the design.

### CODE REVIEW

File: include/myst/regions.h [hunk 1]
```diff
@@ -4,21 +4,77 @@
...
+
+typedef int (*myst_add_page_t)(
+    void* arg,
+    uint64_t vaddr,
+    const void* page,
+    int flags); /* PROT_READ, PROT_WRITE, PROT_EXEC, MYST_REGION_EXTEND */
+
+typedef struct myst_region_context myst_region_context_t;
+
+int myst_region_init(
```
[the remark this answers is not in the document]
Comment (mikbras) [reply include/myst/regions.h:55]: Yes, here: ``tools/myst/host/regions.c``

File: tests/cross_fs_symlinks/.gitignore [hunk 2]
```diff
@@ -0,0 +1,3 @@
+appdir
```
[the remark this answers is not in the document]
Comment (mikbras) [reply tests/cross_fs_symlinks/.gitignore:1]: I want to discuss this with the team. Putting ignored files in the top-level .gitignore is an ongoing source of merge conflicts. LTP places .gitignore in subdirectories. Also I would like to do away with the need for adding files to .gitignore by placing all output under build, at some point anyway.

File: tools/myst/host/regions.c [hunk 3]
```diff
@@ -459,519 +484,228 @@ static int _add_kernel_region(oe_region_context_t* context, uint64_t* vaddr)
...
-                    extend) != OE_OK)
-            {
-                ERAISE(-EINVAL);
-            }
+        memcpy(page, p, min);
 
-            page += PAGE_SIZE;
-            (*vaddr) += PAGE_SIZE;
-        }
+        if (min < sizeof(page))
```
[the remark this answers is not in the document]
Comment (mikbras) [reply tools/myst/host/regions.c]: Nice optimization! Fixed.

File: tools/myst/kargs.c [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+                args->reloc_data,
+                args->reloc_size) != 0)
+        {
+            snprintf(err, err_size, "failed to relocate kernel symbols");
+            ERAISE(-EINVAL);
+        }
+    }
+
+    /* find the kernel symtab region */
+    {
```
[the remark this answers is not in the document]
Comment (mikbras) [reply tools/myst/kargs.c]: Excellent idea. Done.

**Review by jxyang — APPROVED** [review]

**Pull request merged**